src/rpc/mod.rs

```diff
@@ -0,0 +1,106 @@
+use std::{net::SocketAddr, sync::Arc};
+
+use crate::config::Config;
+
+use eyre::Result;
+
+use ethers::{
+    providers::{Middleware, Provider},
+    types::{Block, BlockId, H256},
+    utils::keccak256,
+};
+
+use jsonrpsee::{
+    core::{async_trait, Error},
+    proc_macros::rpc,
+    server::ServerBuilder,
+};
+
+use serde::{Deserialize, Serialize};
+
+#[rpc(server, client, namespace = "optimism")]
+pub trait Rpc {
+    #[method(name = "outputAtBlock")]
+    async fn output_at_block(&self, block_number: u64) -> Result<OutputRootResponse, Error>;
+}
+
+#[derive(Debug)]
+pub struct RpcServerImpl {
+    config: Arc<Config>,
+}
+
+#[async_trait]
+impl RpcServer for RpcServerImpl {
+    async fn output_at_block(&self, block_number: u64) -> Result<OutputRootResponse, Error> {
+        let l2_provider =
+            Provider::try_from(self.config.l2_rpc_url.clone()).expect("invalid L2 RPC url");
```

Comment:
Doing `expect` here will cause rust to panic. Instead we should gracefully handle the error.

[This](https://github.com/a16z/helios/blob/master/client/src/rpc.rs#L333) is a good example of converting any error into an error type that the RPC is happy with. Once we do that, we can just use `?` to automatically return the result if it errors. Take a look at the usage of that `convert_error` function I linked.

Reply:
Such a good patters. Good to learn good practices in rust. Thanks!